dandi/files.py

```diff
@@ -794,26 +801,77 @@ def iter_upload(
                     f"{self.filepath}: Zarr etag changed; was originally"
                     f" {digest['dandi:dandi-zarr-checksum']} but is now {filetag}"
                 )
+        lgr.debug("%s: Producing asset", asset_path)
+        yield {"status": "producing asset"}
+        old_zarr_entries: Dict[str, RemoteZarrEntry] = {}
+        if replacing is not None:
+            lgr.debug("%s: Replacing pre-existing asset")
+            if isinstance(replacing, RemoteZarrAsset):
+                lgr.debug("%s: Pre-existing asset is a Zarr; reusing & updating")
+                zarr_id = replacing.zarr
+                old_zarr_entries = {str(e): e for e in replacing.iterfiles()}
+            else:
+                lgr.debug("%s: Pre-existing asset is not a Zarr; minting new Zarr")
+                r = client.post("/zarr/", json={"name": self.filepath.name})
+                zarr_id = r["zarr_id"]
+            r = client.put(
+                replacing.api_path,
+                json={"metadata": metadata, "zarr_id": zarr_id},
+            )
+        else:
+            lgr.debug("%s: Minting new Zarr")
+            r = client.post("/zarr/", json={"name": self.filepath.name})
+            zarr_id = r["zarr_id"]
+            r = client.post(
+                f"{dandiset.version_api_path}assets/",
+                json={"metadata": metadata, "zarr_id": zarr_id},
+            )
+        a = RemoteAsset.from_data(dandiset, r)
+        to_upload: List[dict] = []
+        for p in stat.files:
+            pdigest = p.get_digest().value
+            item = {"path": str(p), "etag": pdigest}
+            try:
+                e = old_zarr_entries.pop(str(p))
+            except KeyError:
+                to_upload.append(item)
+            else:
+                if e.is_dir():
```

Comment:
having a unittest to cover this codeblock (just delete some folder and/or files within a zarr and reupload?) would be nice

Reply:
Done.